### PR TITLE
Nested NER trainable pipeline component

### PR DESCRIPTION
**Pull request by percevalw**
## Description

The default spacy Named Entity Recognizer (NER) pipeline only allows flat entity
recognition, meaning that overlapping and nested entities cannot be extracted.

The other spacy component `SpanCategorizer` only supports assigning to a specific span
group and it seems that both components are not well suited for extracting entities with ill-defined
boundaries (this can occur if your training data contains difficult and long entities).

This PR introduces a new `nested_ner` component to extract most named entities:
- flat entities like spacy's `EntityRecognizer` overlapping entities
- overlapping entities of different labels (much like spacy's `SpanCategorizer`)

However, the model cannot currently extract entities that are nested inside larger entities
of the same label.

The pipeline assigns both `doc.ents` (in which overlapping entities are filtered
out) and `doc.spans`.

## Checklist

- [x] If this PR is a bug fix, the bug is documented in the test suite.
- [x] Changes were documented in the changelog (pending section).
- [x] If necessary, changes were made to the documentation (eg new pipeline).

### CODE REVIEW

File: edsnlp/utils/training.py [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+        Optional spacy model to load documents from non-spacy formats (like brat)
+    seed: int
+        The seed if we need to shuffle the data when splitting the dataset
+    reader: str
+        Which spacy reader to use when loading the data
+
+    Returns
+    -------
+    Config
+    """
```
Comment (percevalw) [inline edsnlp/utils/training.py:155]: How about this to load the data ? @bdura

File: docs/pipelines/trainable/ner.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
+# Nested Named Entity Recognition
+
+The default spacy Named Entity Recognizer (NER) pipeline only allows flat entity
```
Comment (bdura) [inline docs/pipelines/trainable/ner.md:3]: Can you search mentions of `spacy` and replace them with `spaCy` when needed? Sorry for the chore...

```suggestion
The default spaCy Named Entity Recognizer (NER) pipeline only allows flat entity
```

File: docs/pipelines/trainable/ner.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+# Out: [si folfox inefficace]
+
+# And export new predictions as Brat annotations
+predicted_docs = BratConnector("/path/to/the/new/files", run_pipe=True).brat2docs(nlp)
+BratConnector("/path/to/predictions").docs2brat(predicted_docs)
+```
+
+## Figures
+
+Here are some numbers
```
Comment (bdura) [inline docs/pipelines/trainable/ner.md:88]: A bit of rephrasing here could help 🙂

File: requirements.txt [hunk 4]
```diff
@@ -6,5 +6,7 @@ pendulum
 pydantic>=1.8.2,<1.10.0
 regex<=2022.3.2  # due to dateparser bug issue 1045
 scikit-learn>=1.0.0
-spacy>=3.1<4.0.0
+spacy>=3.1,<4.0.0
+thinc>=8.0.14
+torch
```
Comment (bdura) [inline requirements.txt:11]: This could be an extra dependency, which would dramatically limit the library's footprint for users that are only interested in the rule-based pipelines...

**Review by bdura — COMMENTED** [review]
Great work! I left a few suggestions here and there but this is all very impressive!

File: docs/pipelines/trainable/ner.md (hunk 3 above)
Comment (percevalw) [reply docs/pipelines/trainable/ner.md:88]: Some numbers would too !

File: edsnlp/models/nested_ner.py [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+def flatten_dim(t, dim, ndim=1):
+    return t.reshape((*t.shape[:dim], -1, *t.shape[dim + 1 + ndim :]))
+
+
+class NestedNERModule(torch.nn.Module):
+    def __init__(
+        self,
+        input_size: Optional[int] = None,
+        n_labels: Optional[int] = None,
+        mode: CRFMode = CRFMode.joint,
```
Comment (bdura) [inline edsnlp/models/nested_ner.py:52]: 🎉

File: edsnlp/models/nested_ner.py [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+        n_labels: int
+            Number of labels predicted by the module
+        mode: CRFMode
+            Loss mode of the CRF
+        """
+        super().__init__()
+
+        self.mode = mode
+
+        assert mode in (CRFMode.independent, CRFMode.joint, CRFMode.marginal)
```
Comment (bdura) [inline edsnlp/models/nested_ner.py:70]: I think we might be able to use the Enum constructor for that, although I'm not sure. Or maybe with Pydantic's `validate_arguments`? It's in beta, though.

File: edsnlp/utils/training.py (hunk 1 above)
Comment (bdura) [reply edsnlp/utils/training.py]: Looks good!

**Review by bdura — APPROVED** [review]
